bugbug/similarity.py

```diff
@@ -116,32 +153,23 @@ def get_similar_bugs(self, query, k=10):
         return [self.corpus[j[0]][0] for j in sims[:k]]
 
 
-def evaluation():
-    similarity = TextualSimilarity()
-    total_r = 0
-    hits_r = 0
-    total_p = 0
-    hits_p = 0
-
-    for bug in bugzilla.get_bugs():
-        if duplicates[bug["id"]]:
-            similar_bugs = similarity.get_similar_bugs(bug)
-
-            # Recall
-            for item in duplicates[bug["id"]]:
-                total_r += 1
-                if item in similar_bugs:
-                    hits_r += 1
+class NeighborsSimilarity(BaseSimilarity):
+    def __init__(self, k=10, vectorizer=TfidfVectorizer):
+        self.vectorizer = TfidfVectorizer()
+        self.similarity_calculator = NearestNeighbors(n_neighbors=k)
+        text = []
+        for bug in bugzilla.get_bugs():
+            text.append(text_preprocess(get_text(bug), join=True))
 
-            # Precision
-            for element in similar_bugs:
-                total_p += 1
-                if element in duplicates[bug["id"]]:
-                    hits_p += 1
+        self.vectorizer.fit(text)
+        self.similarity_calculator.fit(self.vectorizer.transform(text))
 
-    print(f"Recall: {hits_r/total_r * 100}%")
-    print(f"Precision: {hits_p/total_p * 100}%")
+    def get_similar_bugs(self, query):
+        bugs = []
+        for bug_data in bugzilla.get_bugs():
+            bugs.append(bug_data["id"])
```

Comment:
Why are you storing data in `bugs` here? It seems like you don't need it?

Reply:
Ah you are using it to get the IDs. You should refactor this, otherwise get_similar_bugs will be extremely slow.

Reply:
E.g. you can make the list of IDs be an attribute of the class, similarly to LSISimilarity (the IDs in that case are stored in the corpus)